### PR TITLE
ARROW-16538: [Java] Adding flexibility to mock ResultSets

### PR DESCRIPTION
**Pull request by toddfarmer**
The minimum required to support existing use cases of FakeResultSet has been implemented here - every other method throws a SQLException, and support can be added in the future as specific methods of MockResultSet are referenced.

### CODE REVIEW

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java [hunk 1]
```diff
@@ -0,0 +1,1617 @@
...
+
+    @Override
+    public boolean isBeforeFirst() throws SQLException {
+      throwIfClosed();
+      return index == 0;
+    }
+
+    @Override
+    public boolean isAfterLast() throws SQLException {
+      return index == rows.size();
```
Comment (lidavidm) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:118]: Given this is the definition of `isLast`, should this be `index > rows.size()`?

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java [hunk 2]
```diff
@@ -0,0 +1,1617 @@
...
+    }
+
+    public static MockResultSetMetaData fromRows(ArrayList<MockRow> rows) throws SQLException {
+
+      if (rows.size() == 0) {
+        throw new SQLException("Unable to dynamically generate ResultSetMetaData because row count is zero!");
+      }
+      MockRow firstRow = rows.get(0);
+      if (firstRow.dataElements.size() == 0) {
+        throw new SQLException("Unable to dynamically generate ResultSetMetaData because column count is zero!");
```
Comment (lidavidm) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:313]: Is it not valid to have a zero-column dataset?

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java [hunk 3]
```diff
@@ -0,0 +1,1617 @@
...
+        return getValueAsString().length();
+      }
+      throw getExceptionToThrow("Unable to determine precision for data type!");
+    }
+
+    private int getScale() throws SQLException {
+      if (this.sqlType == Types.VARCHAR) {
+        return 0;
+      }
+      throw getExceptionToThrow("Unable to determine precision for data type!");
```
Comment (lidavidm) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:417]:
```suggestion
      throw getExceptionToThrow("Unable to determine scale for data type!");
```

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java [hunk 4]
```diff
@@ -0,0 +1,1617 @@
...
+    }
+
+    private int isNullable() throws SQLException {
+      if (this.sqlType == Types.VARCHAR) {
+        return ResultSetMetaData.columnNullable;
+      }
+      throw getExceptionToThrow("Unable to determine precision for data type!");
```
Comment (lidavidm) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:424]: Should this be `columnNullableUnknown`?

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java [hunk 5]
```diff
@@ -0,0 +1,1617 @@
...
+      try {
+        return Boolean.parseBoolean(getValueAsString());
+      } catch (Exception ex) {
+        throw new SQLException(ex);
+      }
+    }
+
+    public int getInt() throws SQLException {
+      try {
+        return Integer.parseInt(getValueAsString());
```
Comment (lidavidm) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:449]: Hmm, why don't we just cast the object we're holding?

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java (hunk 2 above)
Comment (toddfarmer) [reply java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:317]: If it is, it's not possible to dynamically construct a MockResultSetMetaData from a zero-column data set.  I've added a comment to clarify.

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java (hunk 5 above)
Comment (toddfarmer) [reply java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:453]: That's an option, but the thinking behind my approach here was to enable easy type conversion.  By way of example, the following test succeeds with the current code, but will fail with object casting:

```java
  @Test
  public void testMockDataTypes() throws SQLException {
    ResultSetUtility.MockDataElement element = new ResultSetUtility.MockDataElement(1L, Types.NUMERIC);
    assertEquals(1L, element.getLong());
    assertEquals(1, element.getInt());
    assertEquals("1", element.getString());
  }
```

Comment (lidavidm) [reply java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java:455]: Ah, ok, sounds good.

**Review by lidavidm — APPROVED** [review]
LGTM, thank you!

**Pull request closed without merging**